src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -0,0 +1,2168 @@
+	 * `set_attribute` and `add_class`, and an internal state that tracks
+	 * text offsets in the input document.
+	 *
+	 * When higher-level HTML methods are called, those have to transform their
+	 * operations (such as setting an attribute's value) into text diffing
+	 * operations (such as replacing the sub-string from indices A to B with
+	 * some given new string). These text-diffing operations are the lexical
+	 * updates.
+	 *
+	 * As new higher-level methods are added they need to collapse their
+	 * operations into these lower-level lexical updates since that's the
+	 * Tag Processor's internal language of change. Any code which creates
+	 * these lexical updates must ensure that they do not cross HTML syntax
+	 * boundaries, however, so these should never be exposed outside of this
+	 * class or any classes which intentionally expand its functionality.
+	 *
+	 * These are enqueued while editing the document instead of being immediately
+	 * applied to avoid processing overhead, string allocations, and string
+	 * copies when applying many updates to a single document.
+	 *
+	 * Example:
+	 * ```php
+	 *     // Replace an attribute stored with a new value, indices
+	 *     // sourced from the lazily-parsed HTML recognizer.
+	 *     $start = $attributes['src']->start;
+	 *     $end   = $attributes['src']->end;
+	 *     $modifications[] = new WP_HTML_Text_Replacement( $start, $end, $new_value );
+	 *
+	 *     // Correspondingly, something like this will appear in this array.
+	 *     $lexical_updates = [
+	 *         WP_HTML_Text_Replacement( 14, 28, 'https://my-site.my-domain/wp-content/uploads/2014/08/kittens.jpg' )
+	 *     ];
+	 * ```
+	 *
+	 * @since 6.2.0
+	 * @var WP_HTML_Text_Replacement[]
+	 */
+	private $lexical_updates = array();
+
+	/**
+	 * Tracks and limits `seek()` calls to prevent accidental infinite loops.
+	 *
+	 * @see seek
+	 * @since 6.2.0
+	 * @var int
+	 */
+	private $seek_count = 0;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 6.2.0
+	 *
+	 * @param string $html HTML to process.
+	 */
+	public function __construct( $html ) {
+		$this->html = $html;
+	}
+
+	/**
+	 * Finds the next tag matching the $query.
+	 *
+	 * @since 6.2.0
+	 *
+	 * @param array|string|null $query {
+	 *     Optional. Which tag name to find, having which class, etc. Default is to find any tag.
+	 *
+	 *     @type string|null $tag_name     Which tag to find, or `null` for "any tag."
+	 *     @type int|null    $match_offset Find the Nth tag matching all search criteria.
+	 *                                     0 for "first" tag, 2 for "third," etc.
+	 *                                     Defaults to first tag.
+	 *     @type string|null $class_name   Tag must contain this whole class name to match.
+	 *     @type string|null $tag_closers  "visit" or "skip": whether to stop on tag closers, e.g. </div>.
+	 * }
+	 * @return boolean Whether a tag was matched.
+	 */
+	public function next_tag( $query = null ) {
+		$this->parse_query( $query );
```

Comment:
`WP_HTML_Tag_Processor::next_tag` expects `$query` to be a string, an array, or a null.
But it doesn't check its type. This could lead to unpredictable behavior if, for example, an object or a false value is passed to the method (because the method doesn't expect these types).

I **_suggest_** throwing `InvalidArgumentException`, or (this is probably better) calling the `_doing_it_wrong` function to warn developers about the wrong type.

```suggestion
		if ( ! is_null( $query ) && ! is_string( $query ) && ! is_array( $query ) ) {
			throw new InvalidArgumentException('Angry warning message');
			// OR
			_doing_it_wrong( __METHOD__, __( 'Angry warning message' ), '6.2.0' );		
			$query = null; // Probably that's what developer meant
		}		
		
		$this->parse_query( $query );
```

Reply:
Thanks @anton-vlasenko - I'll add a `_doing_it_wrong` call and maybe update the other user warning functions to that as well.

It's not true really that `next_tag` doesn't check the type of the query; it delegates that to `parse_query` which _does_ and so the results won't be unpredictable. If an invalid query is passed it will operate as if a `null` query were passed.

Reply:
Sounds good. If the expected cause is a "developer error" a `_doing_it_wrong()` would be good. If it is a bug, perhaps `InvalidArgumentException` would be better, just needs to be wrapped in `if ( WP_DEBUG )`.

Reply:
they are all `_doing_it_wrong` now